QModManager/Utility/MainMenuMessages.cs

```diff
@@ -0,0 +1,169 @@
+                messages.Clear();
+
+                Patches.Unpatch();
+            }
+        }
+
+        private static class Patches
+        {
+            public static HarmonyInstance hInstance { get; private set; }
+
+            public static void Patch()
+            {
+                Logger.Debug("MainMenuMessages: patching");
+
+                // patching it only if we need to (transpilers take time)
+                hInstance.Patch(AccessTools.Method(typeof(ErrorMessage), nameof(ErrorMessage.OnUpdate)),
+                    transpiler: new HarmonyMethod(AccessTools.Method(typeof(Patches), nameof(UpdateMessages))));
+            }
+
+            public static void Unpatch()
+            {
+                Logger.Debug("MainMenuMessages: unpatching");
+
+                hInstance.Unpatch(AccessTools.Method(typeof(ErrorMessage), nameof(ErrorMessage.Awake)),
+                    AccessTools.Method(typeof(AddMessages), nameof(AddMessages.Postfix)));
+
+                hInstance.Unpatch(AccessTools.Method(typeof(ErrorMessage), nameof(ErrorMessage.OnUpdate)),
+                    AccessTools.Method(typeof(Patches), nameof(UpdateMessages)));
+            }
+
+            [HarmonyPatch(typeof(ErrorMessage), nameof(ErrorMessage.Awake))]
+            private static class AddMessages
+            {
+                // workaround to get harmony instance
+                private static MethodBase TargetMethod(HarmonyInstance instance)
+                {
+                    hInstance = instance;
+                    return null; // using target method from attribute
+                }
+
+                public static void Postfix()
+                {
+                    Logger.Debug($"MainMenuMessages: ErrorMessage created");
+
+                    messageQueue.ForEach(msg => AddInternal(msg));
+                    messageQueue.Clear();
+                }
+            }
+
+            private static float _getVal(float val, ErrorMessage._Message message) => messages.Contains(message)? 1f: val;
+
+            // we changing result for 'float value = Mathf.Clamp01(MathExtensions.EvaluateLine(...' to 1.0f
+            // so text don't stay in the center of the screen (because of changed 'timeEnd')
+            private static IEnumerable<CodeInstruction> UpdateMessages(IEnumerable<CodeInstruction> cins)
+            {
+                var list = new List<CodeInstruction>(cins);
+                int index = list.FindIndex(cin => cin.opcode == OpCodes.Stloc_S && (cin.operand as LocalBuilder)?.LocalIndex == 11);
+
+                list.InsertRange(index, new List<CodeInstruction>
+                {
+                    new CodeInstruction(OpCodes.Ldloc_S, 6),
+                    new CodeInstruction(OpCodes.Call, AccessTools.Method(typeof(Patches), nameof(_getVal)))
+                });
+
+                return list;
+            }
+        }
```

Comment:
All patching in QModManager has been done via attribute patching.
Why not keep it that way?

Reply:
Good question, maybe I'm wrong here.
Applying that particular transpiler takes about 200ms on my machine. I'm ok with applying postfix to `ErrorMessage.Awake` every time (~2ms), but 200ms is imho a bit much for unnecessary code, even for initialization, so it applies only if you have messages.

Reply:
Hm... Okay. Performance is important.